language-plutus-core: add quasiquoter

michaelpj opened the pull request:
This adds a quasiquoter for PLC to `language-plutus-core`. In order to get it scope-safe I had to implement a capture-avoiding substitution function - I've put this in its own module since it seems reusable, and no doubt other people can make it more efficient!

I think this should be useful both for defining terms in `language-plutus-core` itself (e.g. in `Constant/Prelude.hs`), and when generating PLC programs in the PlutusTx compiler.

## Implementation notes

The approach I've taken for antiquotation is to use free object-level variables as metavariables. It's perhaps easiest to see how this works out with an example (taken from the tests):
```
unit = [plcType|(all a (type) (fun a a))|]
one = [plcTerm|(abs a (type) (lam x a x))|]
bool = [plcType|(all a (type) (fun (fun unit a) (fun (fun unit a) a))) |]
true = [plcTerm|(abs a (type) (lam x (fun unit a) (lam y (fun unit a) [x one])))|]
false = [plcTerm|(abs a (type) (lam x (fun unit a) (lam y (fun unit a) [x one])))|]
```
So we can reference `unit`, for example, simply by writing an unbound `unit` type variable. (Yes you will get a type error, albeit not a great one, if you use a type instead of a term and vice versa, and a better error if the referenced variable doesn't exist.)

This is scope-safe as you can see from the test output (which is rendered with `Debug` to verify this). The output for `bool` is:
```
(all a_0 (type) (fun (fun (all a_2 (type) (fun a_2 a_2)) a_0) (fun (fun (all a_2 (type) (fun a_2 a_2)) a_0) a_0)))
```
so even though we've use `a` for all the variables, the uniques have been modified so they don't clash. This is something of a brute-force method, but it seems to work.

The TH is, eventually, quite straightforward. I did quite a lot of head-scratching trying to work out how to do something like `[| captureAvoidingSubstTerm (\n -> $(substFun n)) t |]` - this violates the staging restriction on `n`. In the end, what I came up with was reifying the entire map that encodes the function as a quote, and then building the function back up out of that. Better suggestions welcome!

effectfully commented:
That's an impressive amount of work performed in quite a short time.

Personally, I'm unsatisfied by the fact that we have capture-avoiding substitution at all. It's subtle (though, this particular implementation doesn't seem to be that subtle) and inefficient. I think if we handle names like that, we're doing something wrong.

michaelpj commented:
Agreed 100% on capture-preserving substitution being a pain. However, given the way we currently use names, something like this seems unavoidable (and I've tried to use a fairly sane function interface so y'all can replace it with something better easily).

It's possible that this won't even scale with lots of repeated substitutions either, but it was the only way I could see to get something correct!

effectfully commented:
> However, given the way we currently use names, something like this seems unavoidable

I think we discussed a few weeks ago (before you joined) that a quasiquoter should construct terms in the `Fresh` monad. In the style of

```haskell
getBuiltinChurchSucc :: Fresh (Term TyName Name ())
getBuiltinChurchSucc = do
    nat <- getBuiltinChurchNat
    n <- freshName () "n"
    r <- freshTyName () "r"
    z <- freshName () "z"
    f <- freshName () "f"
    return
        . LamAbs () n nat
        . TyAbs () r (Type ())
        . LamAbs () z (TyVar () r)
        . LamAbs () f (TyFun () (TyVar () r) $ TyVar () r)
        . Apply () (Var () f)
        . foldl (Apply ()) (TyInst () (Var () n) $ TyVar () r)
        $ [ Var () z
          , Var () f
          ]
```

Not that I find it particularly beautiful, but this seems to be an interface we agreed on. @mchakravarty, do I recall this correctly?

> It's possible that this won't even scale with lots of repeated substitutions either, but it was the only way I could see to get something correct!

Yeah, I'm not criticizing your work, it's great.

michaelpj commented:
> I think we discussed a few weeks ago (before you joined) that a quasiquoter should construct terms in the Fresh monad

That makes sense, I think the main thing is that the parser would need to produce its result in `Fresh`, so we could chain the parse of the main term with the parses of previous terms, etc. I also think it's nicer if the values produced are actually just plain `Term`s, which is what this one gets. Then users of the quasiquoter have to use the `dropFresh` thing, which is also pretty unpleasant.

> Yeah, I'm not criticizing your work, it's great.

No offense taken - I'd also love a better approach :)

effectfully commented:
> I also think it's nicer if the values produced are actually just plain `Terms`, which is what this one gets.

That was my argument back then.

> Then users of the quasiquoter have to use the `dropFresh` thing, which is also pretty unpleasant.

Normally, the user is supposed to chain `Fresh` computations and only run the resulting computation at the outermost level using `runFresh`, so no `dropFresh` is required.

vmchale commented on language-plutus-core/src/Language/PlutusCore/Lexer.x:
```diff
@@ -2,6 +2,7 @@
     {-# OPTIONS_GHC -fno-warn-unused-imports #-}
     {-# LANGUAGE DeriveAnyClass     #-}
     {-# LANGUAGE DeriveGeneric      #-}
+    {-# LANGUAGE DeriveLift         #-}
```
If possible, add `DeriveLift` to the `other-extensions` field of the `.cabal` file.

michaelpj replied:
:+1:

vmchale commented on language-plutus-core/src/Language/PlutusCore/Name.hs:
```diff
@@ -23,22 +24,30 @@ import qualified Data.Map             as M
 
+-- derived instance is not okay since it uses the attribute!
+instance Ord (Name a) where
+  n1 <= n2 = (nameString n1) <= (nameString n2) && (nameUnique n1) <= (nameUnique n2)
```
Is there a reason to not simply use `nameUnique` here? That seems like it would be faster.

michaelpj replied:
Ah - is `nameUnique` globally unique? For some reason I thought it was only unique modulo having the same name.

vmchale commented on language-plutus-core/src/Language/PlutusCore/Parser.y:
```diff
@@ -158,7 +162,17 @@ parseST str = liftErr (runAlexST str (runExceptT parsePlutusCore))
 parse :: BSL.ByteString -> Either ParseError (Program TyName Name AlexPosn)
-parse str = liftErr (runAlex str (runExceptT parsePlutusCore))
+parse str = liftErr (runAlex str (runExceptT parsePlutusCoreProgram))
+    where liftErr (Left s)  = Left (LexErr s)
```
might be sensible to pull `liftErr` into a top-level binding to avoid repetition?

michaelpj replied:
:+1:

michaelpj commented:
Some reflection on `Fresh` - I think we probably do want something like this, it's a bit like the `Q` monad in TH. That approach has another benefit: we can control whether our subsitutions are capturing or not. In particular, at the moment, names are *always* distinct between antiquoted terms and the ambient term. But this sometimes isn't what you want when you're building up terms. If you construct things inside `Fresh` you can control that by making fresh identifiers or explicitly passing identifiers that you share between inner and outer.

I'll have a look at reworking this in that direction.

michaelpj commented:
Updated to use `Quote`. The quasiquoter now produces terms in `Quote`, which means they compose nicely. Substitution of metavars is still scope-safe, but doesn't need the capture-avoiding substitution function (I've added a commit deleting that, but it works so we could keep it if we wanted).

The TH is still fairly okay. There is one slightly terrible trick: we need to parse the quasiquoted code twice. We parse it at compile time to determine the metavars so we can generate the appropriate Haskell variable references, and we parse it again at runtime in `QuoteT` so its variables have the appropriate freshness. This is kind of of a hack, but I can't see a way around it since we genuinely need the information at both points.
